src/riverqueue/driver/riversqlalchemy/sql_alchemy_driver.py

```diff
@@ -1,33 +1,63 @@
 from contextlib import contextmanager
-from typing import Optional, List, Generator
+from sqlalchemy import Engine
+from sqlalchemy.engine import Connection
+from typing import Iterator, Optional, List, Generator, cast
 
-from ...driver import DriverProtocol, GetParams, JobInsertParams
+from ...driver import DriverProtocol, ExecutorProtocol, GetParams, JobInsertParams
 from ...model import Job
 from . import river_job, pg_misc
 
 
-class Driver(DriverProtocol):
-    def __init__(self, session):
-        self.session = session
-        self.pg_misc_querier = pg_misc.Querier(session)
-        self.job_querier = river_job.Querier(session)
+class Executor(ExecutorProtocol):
+    def __init__(self, conn: Connection):
+        self.conn = conn
+        self.pg_misc_querier = pg_misc.Querier(conn)
+        self.job_querier = river_job.Querier(conn)
 
     def advisory_lock(self, key: int) -> None:
         self.pg_misc_querier.pg_advisory_xact_lock(key=key)
 
-    def job_insert(self, insert_params: JobInsertParams) -> Optional[Job]:
-        return self.job_querier.job_insert_fast(insert_params)
+    def job_insert(self, insert_params: JobInsertParams) -> Job:
+        return cast(
+            Job,
+            self.job_querier.job_insert_fast(
+                cast(river_job.JobInsertFastParams, insert_params)
+            ),
+        )
 
     def job_insert_many(self, all_params) -> List[Job]:
         raise NotImplementedError("sqlc doesn't implement copy in python yet")
 
     def job_get_by_kind_and_unique_properties(
         self, get_params: GetParams
     ) -> Optional[Job]:
-        return self.job_querier.job_get_by_kind_and_unique_properties(get_params)
+        return cast(
+            Optional[Job],
+            self.job_querier.job_get_by_kind_and_unique_properties(
+                cast(river_job.JobGetByKindAndUniquePropertiesParams, get_params)
+            ),
+        )
 
     @contextmanager
     def transaction(self) -> Generator:
-        session = self.session
-        with session.begin_nested():
-            yield
+        if self.conn.in_transaction():
+            with self.conn.begin_nested():
+                yield
+        else:
+            with self.conn.begin():
+                yield
+
+
+class Driver(DriverProtocol):
+    def __init__(self, conn: Connection | Engine):
+        self.conn = conn
+
+    def get_executor(self) -> Iterator[ExecutorProtocol]:
+        if isinstance(self.conn, Engine):
+            with self.conn.begin() as tx:
+                yield Executor(tx)
```

Comment:
I'm sure you have a good reason for using a generator here that I'm not grokking, but seems odd that the API would return `Iterator[ExecutorProtocol]`.

Reply:
Hm, okay I may have misunderstood something but if you're using a `with` block (on the line above), don't you have to use a generator? If it was just a `return`, the transaction would be closed by the context manager when the function returns, thereby rendering it unusable.

Reply:
You want a context manager instead: https://docs.python.org/3/library/contextlib.html

Reply:
Ah I see. Yeah, the difference between iterators and context managers has been driving me crazy because even if you use a context manager, the returned type is still considered to be an `Iterator` + the `@contextmanager` tag.

I just pushed a change where I attempted to make this a context manager instead. Would you mind taking a look to see if that looks right? It's a separate commit to make the change easier to review in isolation.